server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/GetSnapshotsRequest.java

```diff
@@ -80,7 +86,15 @@ public GetSnapshotsRequest(StreamInput in) throws IOException {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        out.writeString(repository);
+        if (out.getVersion().onOrAfter(MULTIPLE_REPOSITORIES_SUPPORT_ADDED)) {
+            out.writeStringArray(repositories);
+        } else {
+            if (repositories.length != 1) {
```

Comment:
This looks much better now, maybe we should also `assert repositories.length == 1` here though to make tests easier to debug and document that getting into this `if` is clearly a code bug? Otherwise if this ever breaks in REST style tests it's a pain to debug the exact source if it's not an annotation (since the transport layer just catches and logs this exception at best)

Reply:
@original-brownbear I still think this should not be an assertion. Consider the following example. 7.x node receives a request with `_all` repository and forwards it to the current master, which is 8.x node. 8.x node realizes that `_all` is not a single repo and stands for all the repos and produces `GetSnapshotsResponse` containing multiple repositories. 
This `if` will trigger an exception, will be thrown to 7.x node and client will see that multiple repos are not supported.

Reply:
Maybe @ywelsch knows (or has a good idea about it) what the best approach is here when sending a request to an old node that runs against a new master.